runtime/msgpack-serialization.h

```diff
@@ -14,6 +14,57 @@
 #include "runtime/kphp_core.h"
 #include "runtime/string_functions.h"
 
+bool &serialize_as_float32() noexcept;
+
+template<class StreamT, class T>
+inline void pack_value(msgpack::packer<StreamT> &packer, const T& value) {
+  packer.pack(value);
+}
+
+template<class StreamT>
+inline void pack_value(msgpack::packer<StreamT> &packer, double value) {
+  if (serialize_as_float32()) {
+    packer.pack(static_cast<float>(value));
+  } else {
+    packer.pack(value);
+  }
+}
+
+template<class StreamT, class ...Args>
+inline void pack_value(msgpack::packer<StreamT> &packer, const std::tuple<Args...>& value);
+
+template<size_t N>
+struct PackValueHelper {
+  template<class StreamT, class TupleT>
+  static void convert(msgpack::packer<StreamT> &packer, const TupleT &value) {
+    PackValueHelper<N - 1>::convert(packer, value);
+    pack_value(packer, std::get<N - 1>(value));
+  }
+};
+
+template<>
+struct PackValueHelper<0> {
+  template<class StreamT, class TupleT>
+  static void convert(msgpack::packer<StreamT> &, const TupleT &) {}
+};
+
+template<class StreamT, class ...Args>
+inline void pack_value(msgpack::packer<StreamT> &packer, const std::tuple<Args...>& value) {
+  packer.pack_array(sizeof...(Args));
+  PackValueHelper<sizeof...(Args)>::convert(packer, value);
+}
+
+template<class StreamT, class T>
+inline void pack_value(bool as_float32, msgpack::packer<StreamT> &packer, const T& value) {
+  if (as_float32) {
```

Comment:
1) It seems to me, that we have to reset this value in advance, somewhere at packing start, or remove this checking and always set `serialize_as_float32() = as_float32` , so we can be sure, that it would be false by default (in case of SIGALRM right after setting true).

2) How it would work for instances inside instances? If I put float32 tag under another instance, would it pack all floats from it as float32?

Look at this example
```php
<?php


/**
 * @kphp-serializable
 **/
class B {
    /**
     * @kphp-serialized-field 3
     * @kphp-serialized-float32
     * @var double
     */
    public $int_f = 536;
}

/**
 * @kphp-serializable
 **/
class A {
    /**
     * @kphp-serialized-field 3
     * @kphp-serialized-float32
     * @var B
     */
    public $b = null;
}

function run() {
    instance_serialize(new A());
}
run();
```

Reply:
Thanks for the comment, unfortunately, I forgot about signals and fixed up the changes.
`f$instance_serialize` resets the boolean flag for float32 serialization. 

For the second question about instances, the annotation forces everything inside to be encoded as `float32`, you can look at the test case `phpt/msgpack_serialize/023_compound_float32.php`

Reply:
Sorry for bad example(
Please, take a look again
example 1
```php
/** @kphp-serializable **/
class FloatHolder {
    /**
     * @kphp-serialized-field 1
     * @var float[]
     */
    public $float32_f = [20.123, .536, .536];

    /**
     * @kphp-serialized-field 2
     * @var tuple(float, int)
     */
    public $tup;

    public function __construct() {
        $this->tup = tuple(.536, 20);
    }
}

/** @kphp-serializable **/
class A {
    /**
     * @kphp-serialized-field 1
     * @kphp-serialized-float32
     * @var FloatHolder
     */
    public $float_f = null;

    public function __construct() {
        $this->float_f = new FloatHolder();
    }
}
```

example 2 
```php
/** @kphp-serializable **/
class FloatHolder {
    /**
     * @kphp-serialized-field 1
     * @kphp-serialized-float32
     * @var float[]
     */
    public $float32_f = [20.123, .536, .536];

    /**
     * @kphp-serialized-field 2
     * @var tuple(float, int)
     */
    public $tup;

    public function __construct() {
        $this->tup = tuple(.536, 20);
    }
}

/** @kphp-serializable **/
class A {
    /**
     * @kphp-serialized-field 1
     * @kphp-serialized-float32
     * @var FloatHolder
     */
    public $float_f = null;

    public function __construct() {
        $this->float_f = new FloatHolder();
    }
}
```
in the first example `A->float_f->tup` would be serialized as float, in the second example - not.
From my point of view it is confusing

Reply:
Wow, thanks, you are definitely right, it works well on PHP, but fails on kphp.
I've fixed the issue in the new commit (I'll fixup it later).